### PR TITLE
Allow Timer 1 to be instantiated out of phase

### PR DESCRIPTION
**Pull request by mundya**
Add a new method, `set_timer_tick_and_phase`, to allow the timer to be instantiated with a phase offset. This can be used to allow pipelining of the processing stages of some applications.

Example usage project-rig/nengo_spinnaker#134

### CODE REVIEW

**Comment by rowleya** [comment]
Looks like a nice bit of additional functionality.

As an alternative solution to the pipeline problem, would it make more sense to pass a message between the cores i.e. send a multicast message to indicate the second part of the pipeline can now start?  This would be more robust against timers becoming out of sync; even on the same chip, an unexpected overrun of a timer tick could make things go somewhat wrong...

**Comment by mundya** [comment]
> As an alternative solution to the pipeline problem, would it make more sense to pass a message between the cores i.e. send a multicast message to indicate the second part of the pipeline can now start? This would be more robust against timers becoming out of sync; even on the same chip, an unexpected overrun of a timer tick could make things go somewhat wrong...

This is true. For my case timer phase is sufficient (I can largely guarantee the compute time of the process that's triggered by timer tick and am using the phase as a relatively coarse offset), but I agree that for most pipelines it's probably not the best solution.

**Comment by alan-stokes** [comment]
so i've finally had a chance to look at this (now that my testing of the external devices is over). It looks like a clean addition (given its so little lines), but as rowley has said, maybe its worth looking at a more robust fix than changing the spin1_api (esp given the distributed nature of spinnaker) than to try to do funny clock drifts. 

What would happen if the cores using this comm method are on radically different boards and whose clocks have drifted further than 1/2 a timer tick to begin with? are you basically thinking these things are on the same chip and thus has no effect of clock drift? if it is expecting them to be on the same chip, i'm afraid this feels like a hack to exploit your individual application and not really a general case piece of needed functionality, which the api is designed for.

I assume this is related to nengo, and that has much bigger machine usage to us at moment. And so clock drift between boards could be a bigger issue for nengo than us so far, and I don't know if jonathan's work about horses and global clocks ever got into the system, and so you might be buggered before you start if its not on the same chip/board. 

is there any reason why you cant change your code to use a mc command and avoid this api change (i personally like to avoid changing the API as much as possible to be honest)? 

Has @lplana approved these proposed changes to the api?

until these questions have been answered, I'm hesitant to approve it. 
Alan

**Comment by mundya** [comment]
### Should this be in the API?

It doesn't seem unreasonable to want to be able to play with the phase of the timer. Moreover, as you note, this is a minor addition to the code.

> (i personally like to avoid changing the API as much as possible to be honest)

Although not obvious from the diff, this is an addition to the API rather than a change. `spin1_set_timer_tick` is still available and works as before.

> Has @lplana approved these proposed changes to the api?

Luis and I have discussed these changes.

---

### In context

This is used to defer the latter stage of a two-stage pipeline until (most-likely) it has received its input. In general the place-and-route solution will ensure that the stages are closely positioned, but placement on the same board is not guaranteed. Testing of this in use suggests that if clock drift between boards is affecting the precise timing of packet arrival relative to the phased clock tick it does not affect the behaviour of the simulation - which is the metric of choice in Nengo simulations.

Clearly clock drift between boards can be a problem (as, indeed, it already _is_) and I wholeheartedly agree that we should be actively looking to exploit Jonathan's work on synchronisation as soon as possible.

The ideal implementation of what I'm trying to do (delay a computation until input is received) would be best achieved by having the core start processing once the input set is complete. However, as packet delivery is not guaranteed (and some degree of packet loss should be tolerated) another method is needed to trigger the computation. At the moment I'm simply using the phased clock to achieve this, but a later implementation could wait until either the input set is complete or the delayed clock ticks (whichever is sooner) before triggering the second stage of the pipeline.

---

@alan-stokes - I'd suggest that if you want to discuss this further it might be easier to do it in the office tomorrow.

**Comment by neworderofjamie** [comment]
This seem a useful addition to the low-level machine-control API - How it's
used is somewhat application-specific....Saying that though, I THINK it
could be used to get rid of the 2nd timestep of minimum delay when you're
doing split neuron and synapse simulation that @rowleya and I were
discussion.

**Comment by alan-stokes** [comment]
@mundya - discussing in person means that the rest of the community loses the thread. Your welcome to come to my desk and talk, but in the end, someone then needs to update the thread to keep everyone else informed. 

_Although not obvious from the diff, this is an addition to the API rather than a change. spin1_set_timer_tick is still available and works as before._

I understood it didnt change the underlying API calls, but it is still changing the spin1_api and that is what I try to avoid if at all possible. That said, when its been imperative to change it, I have indeed done so, but only when those changes were general and useful enough to warrant such a low level change (the provenance data gathering etc is a prime example, no api changes, but needed much discussion before i was happy to do so).

 _Clearly clock drift between boards can be a problem (as, indeed, it already is) and I wholeheartedly agree that we should be actively looking to exploit Jonathan's work on synchronisation as soon as possible._

this is my biggest hesitation about the proposed change, your opening yourself up to this risk much more (which feels like a bad idea), and therefore as a general support, it feels dangerous. Yet again, given that there's other ways of handling the same thing with solutions that are in use now (reinjection deals with packet loss, other user callback would be my preferred choice that fires off once your input is filled), I'm rather hesitant to put this in just for your use case. I fully concur about jonathans horses. But alas that's not on my task list (i think its a @lplana task list........ correct me if im wrong).

@neworderofjamie  I am not aware of this discussion about 2nd timestep, and thus you'll need to come to my desk to update me. Maybe you should ask both of us to join in your discussions in future? bus factors of 1 are often a bad idea.

@mundya : i appreciate your trying to hack this together before you go to waterloo, but if me and rowley are going to have to support this later on, and if we disagree with the impl, it'll be removed and replaced with what we feel is a software engineered solution. With that in mind, if theres a different road (as we've both implied) then it would be in all our interests for you to go down it now, rather than later. 

you mention that you and luis have discussed it. Im hesitant to approve it till i get @lplana actual opinion, Sergio "discussed" his sdp callback changes with @lplana , and it came to light later on his discussion was one sided. Im not implying you are int he same boat, but having been burnt once...... the whole "trick me once shame on you, trick me twice, shame on me" quote comes to mind. ;-)

anyhow, im spending the next few days clearing out pull requests as much as i can, as there's a massive back log of them, so feel free to pop over if you feel the need. 

regards
Alan

**Comment by lplana** [comment]
I think that the conversation has touched on 3 different topics, which are independent and should be treated separately.

1. (CD) Clock drift between boards.
2. (PD) Phase difference between timers on different cores.
3. (PS) "Core" pipeline synchronization.

1. CD needs to be addressed. Johnny (@mossblaser) did very good work, and picking it up is in my TODO list, albeit with low priority. There hasn't really been a need for it yet.

2. PD is not a problem that needs solving, it's a feature that can be offered. It is different from CD. In fact, it is affected by CD, so it will benefit from any fix to that problem. Fixing CD would make PD more reliable, hence more usable.

3. PS, again, is different from the two above. It came up in this conversation because, if I understand correctly, Andrew (@mundya) is using PD as a simple mechanism to do PS. Alternatives have been proposed, such as using MC packets, and this is a choice that application programmers have to make.

In summary, I think that the central point for this conversation is the value of PD as a feature to be offered and I think that it adds value with little cost. And it's backwards compatible.

I've reviewed the code changes and I'm happy for them to be merged into master.

**Comment by rowleya** [comment]
I am happy with this as a feature of the system - how it is used is up to the user!

On the subject of use, this is an easy way to add in a random phase shift between cores to prevent too much synchronisation between sending of packets of cores on the same chip.

**Comment by mundya** [comment]
The consensus appears to be that this can be merged, would it be possible to do this please? @lplana is assigned.

Thank you :)

**Comment by rowleya** [comment]
I just wanted to do a couple of tests to make sure that this doesn't have any knock-on effects on other scripts.  I will try to do these today...

**Comment by mundya** [comment]
Thanks, there's no particular rush other than that this has sat open for some time.

**Comment by rowleya** [comment]
Tested and seems to cause no issues, so merging

**Comment by mundya** [comment]
Thanks!
